### ISSUE TITLE
[Flight] Upgrade to tonic > 0.3.1 to get fix of headers/trailers handling

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#140]
*Note*: migrated from original JIRA: https://issues.apache.org/jira/browse/ARROW-10961

C++ gRPC servers sometimes return both headers and trailers, and [tonic|https://crates.io/crates/tonic], the crate that provides a Rust gRPC implementation, wasn't correctly merging headers and trailers for errors in the gRPC client. [This has been fixed|https://github.com/hyperium/tonic/pull/510] and should be included in the next release of tonic, which should have some version number greater than 0.3.1, but I'm not sure what tonic's release plans are.

In the Rust Flight integration test client I'm developing, the middleware scenario with the Rust client against the C++ server will fail until this is taken care of. Filing this ticket so I can reference it in the disabling of that test case.

**Comment by lidavidm** [comment]
FWIW, I happened to see this. It looks like we're on tonic 0.8 now - so maybe this can be re-enabled?

**Comment by alamb** [comment]
I agree -- @carols10cents  since it appears you filed the original issue, do you remember which integration scenario you were referring to in the original ticket?

**Comment by lidavidm** [comment]
Sorry, should've linked it: https://github.com/apache/arrow/blob/74f221c925688a1fab05c0394818256d816ccfc1/dev/archery/archery/integration/runner.py#L430-L434

I saw it while poking around the tests

**Comment by carols10cents** [comment]
> I agree -- @carols10cents since it appears you filed the original issue, do you remember which integration scenario you were referring to in the original ticket?

Sorry, I have trouble remembering last week sometimes, let alone December 2020 😅

**Comment by lidavidm** [comment]
I linked it above - I can just re-enable it and see what happens, and if it works we can close this out?

**Comment by alamb** [comment]
>  I linked it above - I can just re-enable it and see what happens, and if it works we can close this out?

That would be awesome @lidavidm  -- thank you very much!

**Comment by lidavidm** [comment]
Looks like it passes fine: https://github.com/apache/arrow/pull/13852
